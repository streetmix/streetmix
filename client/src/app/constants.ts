export const JUST_SIGNED_IN_PATH = '/services/auth/just-signed-in'

// Path segments
export const URL_NEW_STREET = '/new'
export const URL_NEW_STREET_COPY_LAST = '/copy-last'
export const URL_GLOBAL_GALLERY = '/gallery'
export const URL_ERROR = '/error'
export const URL_HELP = '/help'
export const URL_SURVEY_FINISHED = '/survey-finished'

// Error fragments that occur after /error/
export const URL_ERROR_NO_TWITTER_REQUEST_TOKEN = 'no-twitter-request-token'
export const URL_ERROR_NO_TWITTER_ACCESS_TOKEN = 'no-twitter-access-token'
export const URL_ERROR_NO_ACCESS_TOKEN = 'no-access-token'
export const URL_ERROR_AUTHENTICATION_API_PROBLEM = 'authentication-api-problem'
export const URL_ERROR_ACCESS_DENIED = 'access-denied'

export const URL_EXAMPLE_STREET = '/streetmix/7'

// Since URLs like “streetmix.net/new” are reserved, but we still want
// @new to be able to use Streetmix, we prefix any reserved URLs with ~
export const RESERVED_URLS = [
  '/services',
  URL_NEW_STREET,
  URL_NEW_STREET_COPY_LAST,
  URL_GLOBAL_GALLERY,
  URL_ERROR,
  URL_HELP,
  URL_SURVEY_FINISHED,
  '/streets',
  '/terms-of-service',
  '/privacy-policy',
  '/map',
  '/survey'
]

export const URL_RESERVED_PREFIX = '~'

// Color modes
export const COLOR_MODE_LIGHT = 'light'
export const COLOR_MODE_DARK = 'dark'
export const COLOR_MODE_AUTO = 'auto'

export type ColorModes = 'light' | 'dark' | 'auto'
