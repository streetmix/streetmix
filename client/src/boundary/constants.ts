export const BOUNDARY_SPACE = 360
export const BOUNDARY_LEFT_POSITION = 'left'
export const BOUNDARY_RIGHT_POSITION = 'right'

export const MAX_BUILDING_FLOORS = 20
