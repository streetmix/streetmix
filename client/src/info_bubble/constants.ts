export const INFO_BUBBLE_TYPE_SEGMENT = 1
export const INFO_BUBBLE_TYPE_LEFT_BUILDING = 2
export const INFO_BUBBLE_TYPE_RIGHT_BUILDING = 3
