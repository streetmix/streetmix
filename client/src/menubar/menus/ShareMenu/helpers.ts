import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
