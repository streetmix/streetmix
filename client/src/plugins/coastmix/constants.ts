export const SEA_LEVEL_YEAR_2030 = 2030
export const SEA_LEVEL_YEAR_2050 = 2050
export const SEA_LEVEL_YEAR_2070 = 2070
