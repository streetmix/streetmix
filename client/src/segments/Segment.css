/* TODO: Further split-up and refactoring of styles */

.segment {
  display: block;
  position: absolute;
  top: calc(-1 * var(--canvas-ground));
  height: var(--canvas-height);
  text-align: center;
  perspective: 500px;
  touch-action: none;
}

.segment-canvas-container {
  width: 100%;
  height: 100%;
}

body.segment-move-dragging .segment-empty {
  display: none;
}

body:not(.segment-resize-dragging) .segment {
  transition:
    width 150ms,
    left 150ms,
    transform 150ms;
}

body.immediate-segment-resize .segment {
  transition: none !important;
}

.segment.active {
  z-index: var(--z-02-segment-focused) !important;
}

.segment.active .active-bg {
  background: var(--segment-active-background);
  position: absolute;
  left: 0;
  right: 0;
  top: 10px;
  z-index: -1;
  height: calc(var(--canvas-baseline) - 10px);
  pointer-events: none;
}

body:not(.read-only) .segment {
  cursor: grab;
}

.segment.dragged-out {
  opacity: 0;
}

body.segment-move-dragging,
body.segment-move-dragging * {
  cursor: grabbing !important;
}

body.not-within-canvas,
body.not-within-canvas * {
  cursor: no-drop !important;
}

/* NOTE: targeting immediate child div because CSSTransition wrapper workaround
   -- only on removal transition, not switching transition */
.slice-remove-exit > div,
.switching-away-exit {
  position: absolute;
  z-index: var(--z-10-switch-away);
  pointer-events: none;
  transform: none !important;
  transition: none !important;
  perspective: 400px;
}

.slice-remove-exit-done > div,
.switching-away-exit-done {
  z-index: var(--z-10-switch-away) !important;
}

.slice-remove-exit-active canvas,
.switching-away-exit-active canvas {
  transition:
    transform var(--segment-switching-time) ease-in,
    opacity var(--segment-switching-time) ease-in !important;
  transform: rotateX(-60deg);
  transform-origin: 50% 120%;
  opacity: 0;
}

.switching-in-enter {
  z-index: -1 !important;
  perspective: 400px;

  canvas {
    transform: rotateX(90deg);
    transform-origin: 50% 100%;
    opacity: 0;
  }
}

.switching-in-enter-done {
  z-index: -1 !important;
}

.switching-in-enter-active canvas {
  opacity: 1;
  transition:
    transform var(--segment-switching-time),
    opacity var(--segment-switching-time) !important;
  transform: none;
}
