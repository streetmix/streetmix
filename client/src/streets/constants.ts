export const NEW_STREET_DEFAULT = 1
export const NEW_STREET_EMPTY = 2

export const MIN_CUSTOM_STREET_WIDTH = 3 // meters
export const MAX_CUSTOM_STREET_WIDTH = 120 // meters
export const MIN_CUSTOM_STREET_WIDTH_IMPERIAL = 10 * 0.3048 // feet
export const MAX_CUSTOM_STREET_WIDTH_IMPERIAL = 400 * 0.3048 // feet

export const DEFAULT_CAPACITY_SOURCE = 'giz'
