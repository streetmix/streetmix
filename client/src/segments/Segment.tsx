import React, { useState, useRef, useEffect, useCallback } from 'react'
import { CSSTransition } from 'react-transition-group'
import { useDrag, useDrop } from 'react-dnd'

import { useSelector, useDispatch } from '~/src/store/hooks'
import EmptyDragPreview from '~/src/ui/dnd/EmptyDragPreview'
import { usePrevious } from '~/src/util/usePrevious'
import { infoBubble } from '../info_bubble/info_bubble'
import { INFO_BUBBLE_TYPE_SEGMENT } from '../info_bubble/constants'
import { formatMessage } from '../locales/locale'
import { setActiveSegment } from '../store/slices/ui'
import {
  incrementSegmentWidth,
  removeSegmentAction,
  clearSegmentsAction
} from '../store/actions/street'
import { addToast } from '../store/slices/toasts'
import { getSegmentCapacity } from './capacity'
import { getLocaleSegmentName } from './view'
import SegmentCanvas from './SegmentCanvas'
import SegmentDragHandles from './SegmentDragHandles'
import SegmentLabelContainer from './SegmentLabelContainer'

import {
  TILE_SIZE,
  SEGMENT_WARNING_OUTSIDE,
  SEGMENT_WARNING_WIDTH_TOO_SMALL,
  SEGMENT_WARNING_WIDTH_TOO_LARGE
} from './constants'
import {
  createSliceDragSpec,
  createSliceDropTargetSpec
} from './drag_and_drop'
import { getSegmentInfo } from './info'
import { RESIZE_TYPE_INCREMENT } from './resizing'
import type { SliceItem, UnitsSetting } from '@streetmix/types'
import './Segment.css'

interface SliceProps {
  sliceIndex: number
  segment: SliceItem
  units: UnitsSetting
  segmentLeft: number
}

function Segment (props: SliceProps): React.ReactNode {
  const { sliceIndex, segment, units, segmentLeft } = props
  const [switchSegments, setSwitchSegments] = useState(false)
  const [oldVariant, setOldVariant] = useState<string>(segment.variantString)

  const enableAnalytics = useSelector(
    (state) => state.flags.ANALYTICS.value && state.street.showAnalytics
  )
  const locale = useSelector((state) => state.locale.locale)
  const descriptionVisible = useSelector(
    (state) => state.infoBubble.descriptionVisible
  )
  const activeSegment = useSelector((state) =>
    typeof state.ui.activeSegment === 'number' ? state.ui.activeSegment : null
  )
  const capacitySource = useSelector((state) => state.street.capacitySource)
  const dispatch = useDispatch()

  // What is this?
  const initialRender = useRef(true)
  const streetSegment = useRef<HTMLDivElement>(null)
  const dndRef = useRef<HTMLDivElement>(null)

  // These refs are a workaround for CSSTransition's dependence on
  // findDOMNode, which is deprecated.
  const oldRef = useRef<HTMLDivElement>(null)
  const newRef = useRef<HTMLDivElement>(null)

  // Set up drag and drop targets
  // Specs are created on each render with changed props
  const dropSpec = createSliceDropTargetSpec(props, streetSegment)
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [collectedProps, drop] = useDrop(dropSpec)
  const dragSpec = createSliceDragSpec(props)
  const [collected, drag, dragPreview] = useDrag(dragSpec)
  const { isDragging }: { isDragging: boolean } = collected
  drag(drop(dndRef))

  // Keep previous state for comparisons (ported from legacy behavior)
  const prevProps = usePrevious({
    segment,
    isDragging
  })

  useEffect(() => {
    // TODO: there should be checks if the calls to the prop methods should be made in the first place. see discussion here: https://github.com/streetmix/streetmix/pull/1227#discussion_r263536187
    // During a segment removal or a dragging action, the infoBubble temporarily does not appear
    // for the hovered/dragged segment. Once the removal or drag action ends, the infoBubble for
    // the active segment should be shown. The following IF statement checks to see if a removal
    // or drag action occurred previously to this segment and displays the infoBubble for the
    // segment if it is equal to the activeSegment and no infoBubble was shown already.
    if (prevProps === undefined) return

    const wasDragging =
      (prevProps.isDragging && !isDragging) ||
      (initialRender.current && activeSegment !== null)

    initialRender.current = false

    if (wasDragging && activeSegment === sliceIndex) {
      infoBubble.considerShowing(
        false,
        streetSegment.current,
        INFO_BUBBLE_TYPE_SEGMENT
      )
    }
  }, [isDragging, activeSegment, sliceIndex])

  useEffect(() => {
    if (
      prevProps !== undefined &&
      prevProps.segment.variantString !== segment.variantString
    ) {
      handleSwitchSegments(prevProps.segment.variantString)
    }
  }, [segment.variantString])

  // Also animate the switching if elevation changes.
  // Maybe we don't always do this forever, but it makes it match
  // existing elevation variant behavior
  useEffect(() => {
    if (
      prevProps !== undefined &&
      prevProps.segment.elevation !== segment.elevation
    ) {
      handleSwitchSegments(prevProps.segment.variantString)
    }
  }, [segment.elevation])

  // Cleanup effect
  useEffect(() => {
    // Event handler is only added on mouseover, but definitely remove if
    // component is unmounted.
    return () => {
      document.removeEventListener('keydown', handleKeyDown)
    }
  }, [])

  // When called by CSSTransition `onExited`, `oldVariant` is not passed to the
  // function (is undefined). `switchSegments` should be `true` when this happens.
  function handleSwitchSegments (oldVariant?: string): void {
    setSwitchSegments(!switchSegments)
    if (switchSegments) {
      setOldVariant(segment.variantString)
    } else {
      if (oldVariant === undefined) {
        throw new Error('oldVariant should be defined')
      }
      setOldVariant(oldVariant)
    }
  }

  function handleSegmentMouseEnter (event: React.MouseEvent): void {
    dispatch(setActiveSegment(sliceIndex))

    document.addEventListener('keydown', handleKeyDown)
    infoBubble.considerShowing(
      event,
      streetSegment.current,
      INFO_BUBBLE_TYPE_SEGMENT
    )
  }

  function handleSegmentMouseLeave (): void {
    document.removeEventListener('keydown', handleKeyDown)
    infoBubble.dontConsiderShowing()
  }

  function decrementWidth (position: number, finetune: boolean): void {
    void dispatch(
      incrementSegmentWidth(
        position, // slice index
        false, // subtract
        finetune, // true if shift key is pressed
        RESIZE_TYPE_INCREMENT
      )
    )
  }

  function incrementWidth (position: number, finetune: boolean): void {
    void dispatch(
      incrementSegmentWidth(
        position, // slice index
        true, // add
        finetune, // true if shift key is pressed
        RESIZE_TYPE_INCREMENT
      )
    )
  }

  // `event` type is not a React event because listener is attached through DOM
  // We need to define a callback so React can properly clean up event handlers
  const handleKeyDown = useCallback((event: KeyboardEvent): void => {
    switch (event.key) {
      case '-':
      case '_':
        if (event.metaKey || event.ctrlKey || event.altKey) return

        event.preventDefault()
        decrementWidth(sliceIndex, event.shiftKey)
        break
      // Plus (+) may only triggered with shift key, so also check if
      // the same physical key (Equal) is pressed
      case '+':
      case '=':
        if (event.metaKey || event.ctrlKey || event.altKey) return

        event.preventDefault()
        incrementWidth(sliceIndex, event.shiftKey)
        break
      case 'Backspace':
      case 'Delete':
        // Prevent deletion from occurring if the description is visible
        if (descriptionVisible) return

        // If the shift key is pressed, we remove all segments
        if (event.shiftKey) {
          void dispatch(clearSegmentsAction())
          infoBubble.hide()
          dispatch(
            addToast({
              message: formatMessage(
                'toast.all-segments-deleted',
                'All segments have been removed.'
              ),
              component: 'TOAST_UNDO'
            })
          )
        } else {
          infoBubble.hide()
          infoBubble.hideSegment()
          dispatch(
            addToast({
              message: formatMessage(
                'toast.segment-deleted',
                'The segment has been removed.'
              ),
              component: 'TOAST_UNDO'
            })
          )
          void dispatch(removeSegmentAction(sliceIndex))
        }
        break
      default:
        break
    }
  }, [])

  function renderSegmentCanvas (
    variantType: string,
    nodeRef: React.RefObject<HTMLDivElement>
  ): React.ReactNode {
    const isOldVariant = variantType === 'old'

    // The segment ID is a string that uniquely identifies the segment
    // and can be used as a consistent and reliable seed for a PRNG
    const randSeed = segment.id

    return (
      <div ref={nodeRef} style={{ width: '100%', height: '100%' }}>
        <SegmentCanvas
          actualWidth={segment.width}
          type={segment.type}
          variantString={isOldVariant ? oldVariant : segment.variantString}
          randSeed={randSeed}
          elevation={segment.elevation}
        />
      </div>
    )
  }

  const segmentInfo = getSegmentInfo(segment.type)

  // Get localized names from store, fall back to segment default names if
  // translated text is not found. TODO: port to react-intl/formatMessage later.
  const displayName =
    segment.label ?? getLocaleSegmentName(segment.type, segment.variantString)

  const average = getSegmentCapacity(segment, capacitySource)?.average
  const elementWidth = segment.width * TILE_SIZE

  const segmentStyle = {
    width: elementWidth + 'px',
    zIndex: segmentInfo.zIndex,
    transform: `translateX(${segmentLeft}px)`
  }

  const classNames = ['segment']

  if (isDragging) {
    classNames.push('dragged-out')
  } else if (activeSegment === sliceIndex) {
    classNames.push('hover', 'show-drag-handles')
  }

  // Warnings
  if (
    segment.warnings[SEGMENT_WARNING_OUTSIDE] ||
    segment.warnings[SEGMENT_WARNING_WIDTH_TOO_SMALL] ||
    segment.warnings[SEGMENT_WARNING_WIDTH_TOO_LARGE]
  ) {
    classNames.push('warning')
  }
  if (segment.warnings[SEGMENT_WARNING_OUTSIDE]) {
    classNames.push('outside')
  }

  return (
    <div
      style={segmentStyle}
      className={classNames.join(' ')}
      data-testid="segment"
      data-slice-index={sliceIndex}
      ref={streetSegment}
      onMouseEnter={handleSegmentMouseEnter}
      onMouseLeave={handleSegmentMouseLeave}
    >
      <SegmentLabelContainer
        label={displayName}
        width={segment.width}
        units={units}
        locale={locale}
        capacity={average}
        showCapacity={enableAnalytics}
      />
      <SegmentDragHandles width={elementWidth} />
      <div ref={dndRef} className="segment-canvas-container">
        <CSSTransition
          key="old-variant"
          in={!switchSegments}
          classNames="switching-away"
          timeout={250}
          onExited={handleSwitchSegments}
          unmountOnExit={true}
          nodeRef={oldRef}
        >
          {renderSegmentCanvas('old', oldRef)}
        </CSSTransition>
        <CSSTransition
          key="new-variant"
          in={switchSegments}
          classNames="switching-in"
          timeout={250}
          unmountOnExit={true}
          nodeRef={newRef}
        >
          {renderSegmentCanvas('new', newRef)}
        </CSSTransition>
      </div>
      <div className="hover-bk" />
      <EmptyDragPreview dragPreview={dragPreview} />
    </div>
  )
}

export default Segment
