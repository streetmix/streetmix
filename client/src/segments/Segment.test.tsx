import React from 'react'
import { vi } from 'vitest'
import { userEvent } from '@testing-library/user-event'

import { render, screen } from '~/test/helpers/render'
import { infoBubble } from '~/src/info_bubble/info_bubble'
import { setLastStreet } from '~/src/streets/data_model'
import { SETTINGS_UNITS_METRIC } from '~/src/users/constants'
import Segment from './Segment'
import type { SliceItem } from '@streetmix/types'

vi.mock('../info_bubble/info_bubble')

describe('Segment', () => {
  const increment = 0.05
  let variantString, type, activeElement: number, segment: SliceItem

  beforeEach(() => {
    variantString = 'inbound|regular'
    type = 'streetcar'
    activeElement = 0
    segment = {
      type,
      variantString,
      variant: {
        direction: 'inbound',
        'public-transit-asphalt': 'regular'
      },
      id: '1',
      width: 5,
      elevation: 0,
      warnings: []
    }
  })

  it('renders correctly', () => {
    const { asFragment } = render(
      <Segment
        segment={segment}
        sliceIndex={activeElement}
        segmentLeft={0}
        units={SETTINGS_UNITS_METRIC}
      />,
      {
        initialState: {
          flags: { ANALYTICS: { value: true }, DEBUG_SEGMENT_CANVAS_RECTANGLES: { value: false } },
          ui: { activeSegment: activeElement },
          street: { showAnalytics: true, segments: [segment] }
        }
      }
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('shows the info bubble on mouseover', async () => {
    const user = userEvent.setup()

    render(
      <Segment
        segment={segment}
        sliceIndex={activeElement}
        segmentLeft={0}
        units={SETTINGS_UNITS_METRIC}
      />,
      {
        initialState: {
          ui: { activeSegment: activeElement },
          street: { segments: [segment] }
        }
      }
    )

    await user.hover(screen.getByTestId('segment'))

    expect(infoBubble.considerShowing).toHaveBeenCalled()
  })

  it('hides the info bubble on mouseleave', async () => {
    const user = userEvent.setup()

    render(
      <Segment
        segment={segment}
        sliceIndex={activeElement}
        segmentLeft={0}
        units={SETTINGS_UNITS_METRIC}
      />,
      {
        initialState: {
          ui: { activeSegment: activeElement },
          street: { segments: [segment] }
        }
      }
    )

    await user.hover(screen.getByTestId('segment'))
    await user.unhover(screen.getByTestId('segment'))

    expect(infoBubble.dontConsiderShowing).toHaveBeenCalledTimes(1)
  })

  describe('keyboard events', () => {
    it('decreases the width of the segment when minus key is pressed', async () => {
      const user = userEvent.setup()
      const { store } = render(
        <Segment
          segment={segment}
          sliceIndex={activeElement}
          segmentLeft={0}
          units={SETTINGS_UNITS_METRIC}
        />,
        {
          initialState: {
            ui: { activeSegment: activeElement },
            street: { segments: [segment] }
          }
        }
      )

      await user.hover(screen.getByTestId('segment'))
      await user.keyboard('-')

      expect(store.getState().street.segments[activeElement].width).toEqual(
        segment.width - increment
      )
    })

    it('increases the width of the segment when plus key is pressed', async () => {
      const user = userEvent.setup()
      const { store } = render(
        <Segment
          segment={segment}
          sliceIndex={activeElement}
          segmentLeft={0}
          units={SETTINGS_UNITS_METRIC}
        />,
        {
          initialState: {
            ui: { activeSegment: activeElement },
            street: { segments: [segment] }
          }
        }
      )

      await user.hover(screen.getByTestId('segment'))
      await user.keyboard('+')

      expect(store.getState().street.segments[activeElement].width).toEqual(
        segment.width + increment
      )
    })

    it('removes segment when delete key is pressed', async () => {
      const user = userEvent.setup()
      const { store } = render(
        <Segment
          segment={segment}
          sliceIndex={activeElement}
          segmentLeft={0}
          units={SETTINGS_UNITS_METRIC}
        />,
        {
          initialState: {
            ui: { activeSegment: activeElement },
            street: { segments: [segment] }
          }
        }
      )

      setLastStreet() // ToDo: needs to be refactored

      await user.hover(screen.getByTestId('segment'))
      await user.keyboard('{Delete}')

      expect(infoBubble.hide).toHaveBeenCalledTimes(1)
      expect(infoBubble.hideSegment).toHaveBeenCalledTimes(1)
      expect(store.getState().street.segments.length).toEqual(0)
    })

    it('removes all segments when shift+delete keys are pressed', async () => {
      const user = userEvent.setup()
      const { store } = render(
        <Segment
          segment={segment}
          sliceIndex={activeElement}
          segmentLeft={0}
          units={SETTINGS_UNITS_METRIC}
        />,
        {
          initialState: {
            ui: { activeSegment: activeElement },
            street: { segments: [segment] }
          }
        }
      )

      await user.hover(screen.getByTestId('segment'))
      await user.keyboard('{Shift>}{Delete}{/Shift}')

      expect(store.getState().street.segments.length).toEqual(0)
    })
  })
})
