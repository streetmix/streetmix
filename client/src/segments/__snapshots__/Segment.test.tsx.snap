// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Segment > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="segment hover show-drag-handles"
    data-slice-index="0"
    data-testid="segment"
    style="width: 196.8503937007874px; z-index: 15; transform: translateX(0px);"
  >
    <div
      class="segment-label-container"
    >
      <div
        class="segment-grid units-metric"
      />
      <div
        class="segment-width"
      >
        <span>
          5 m
        </span>
      </div>
      <div
        class="segment-label"
      >
        <p
          class="segment-name"
        >
          Streetcar
        </p>
        <p
          class="segment-capacity"
        >
          18,000 people/hr
        </p>
      </div>
    </div>
    <span
      class="drag-handle drag-handle-left"
    >
      <svg
        data-icon="chevron-left"
        data-icon-source="test"
      />
    </span>
    <span
      class="drag-handle drag-handle-right"
    >
      <svg
        data-icon="chevron-right"
        data-icon-source="test"
      />
    </span>
    <div
      class="segment-canvas-container"
    >
      <div
        style="width: 100%; height: 100%;"
      >
        <canvas
          class="segment-image"
          height="445"
          style="width: 197px; height: 445px; left: 0px;"
          width="197"
        />
      </div>
    </div>
    <div
      class="hover-bk"
    />
    <img
      alt=""
      src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
      style="display: none;"
    />
  </div>
</DocumentFragment>
`;
